From 3da34c1eb522fcd728cc440ad5d7ba92e799b545 Mon Sep 17 00:00:00 2001
From: Alain Volmat <alain.volmat@foss.st.com>
Date: Mon, 10 Feb 2025 13:53:22 +0100
Subject: [PATCH] media: dcmipp: correct dcmipp min/max frame size handling

The DCMIPP is able to handle frames up to 4094x4094 with minimum
frame size of 16x2 pixels.  Since dump pipe width is written in
words, possible width can actually be up to 4 * 4094 pixels wide.
By disabling the crop block it is also possible to reach higher
frame size.

Change-Id: Ic07cb37b0c86df80e996e5640347cf2008b9df5d
Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/436362
Tested-by: Hugues FRUCHET <hugues.fruchet@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Hugues FRUCHET <hugues.fruchet@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Hugues FRUCHET <hugues.fruchet@st.com>
---
 .../st/stm32/stm32-dcmipp/dcmipp-byteproc.c      |  4 ++--
 .../st/stm32/stm32-dcmipp/dcmipp-common.h        | 16 +++++++++++++---
 .../st/stm32/stm32-dcmipp/dcmipp-input.c         |  4 ++--
 .../platform/st/stm32/stm32-dcmipp/dcmipp-isp.c  |  8 ++++----
 .../st/stm32/stm32-dcmipp/dcmipp-pixelcap.c      |  8 ++++----
 .../st/stm32/stm32-dcmipp/dcmipp-pixelproc.c     | 12 ++++++------
 6 files changed, 31 insertions(+), 21 deletions(-)

--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-byteproc.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-byteproc.c
@@ -219,9 +219,9 @@ static void dcmipp_byteproc_adjust_fmt_c
 	}
 
 	fmt->width = clamp_t(u32, fmt->width, DCMIPP_FRAME_MIN_WIDTH,
-			     DCMIPP_FRAME_MAX_WIDTH) & ~1;
+			     DCMIPP_FRAME_MAX_WIDTH);
 	fmt->height = clamp_t(u32, fmt->height, DCMIPP_FRAME_MIN_HEIGHT,
-			      DCMIPP_FRAME_MAX_HEIGHT) & ~1;
+			      DCMIPP_FRAME_MAX_HEIGHT);
 
 	if (fmt->field == V4L2_FIELD_ANY || fmt->field == V4L2_FIELD_ALTERNATE)
 		fmt->field = fmt_default.field;
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-common.h
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-common.h
@@ -20,10 +20,20 @@
 
 #define DCMIPP_PDEV_NAME "dcmipp"
 
-#define DCMIPP_FRAME_MAX_WIDTH 4096
-#define DCMIPP_FRAME_MAX_HEIGHT 2160
 #define DCMIPP_FRAME_MIN_WIDTH 16
-#define DCMIPP_FRAME_MIN_HEIGHT 16
+#define DCMIPP_FRAME_MIN_HEIGHT 2
+
+/*
+ * Maximum frame size for input & byte pipe
+ * This is aligned with TPG maximum frame size,
+ * aligned to word access for the width
+ */
+#define DCMIPP_FRAME_MAX_WIDTH	16376
+#define DCMIPP_FRAME_MAX_HEIGHT	16383
+
+/* Maximum frame size for pixel pipes */
+#define DCMIPP_PIXEL_FRAME_MAX_WIDTH 4094
+#define DCMIPP_PIXEL_FRAME_MAX_HEIGHT 4094
 
 #define DCMIPP_FMT_WIDTH_DEFAULT  640
 #define DCMIPP_FMT_HEIGHT_DEFAULT 480
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-input.c
@@ -269,9 +269,9 @@ static void dcmipp_inp_adjust_fmt(struct
 		fmt->code = fmt_default.code;
 
 	fmt->width = clamp_t(u32, fmt->width, DCMIPP_FRAME_MIN_WIDTH,
-			     DCMIPP_FRAME_MAX_WIDTH) & ~1;
+			     DCMIPP_FRAME_MAX_WIDTH);
 	fmt->height = clamp_t(u32, fmt->height, DCMIPP_FRAME_MIN_HEIGHT,
-			      DCMIPP_FRAME_MAX_HEIGHT) & ~1;
+			      DCMIPP_FRAME_MAX_HEIGHT);
 
 	if (fmt->field == V4L2_FIELD_ANY || fmt->field == V4L2_FIELD_ALTERNATE)
 		fmt->field = fmt_default.field;
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-isp.c
@@ -170,9 +170,9 @@ static void dcmipp_isp_adjust_fmt(struct
 					  ISP_MEDIA_BUS_SINK_FMT_DEFAULT;
 
 	fmt->width = clamp_t(u32, fmt->width, DCMIPP_FRAME_MIN_WIDTH,
-			     DCMIPP_FRAME_MAX_WIDTH) & ~1;
+			     DCMIPP_PIXEL_FRAME_MAX_WIDTH) & ~1;
 	fmt->height = clamp_t(u32, fmt->height, DCMIPP_FRAME_MIN_HEIGHT,
-			      DCMIPP_FRAME_MAX_HEIGHT);
+			      DCMIPP_PIXEL_FRAME_MAX_HEIGHT);
 
 	if (fmt->field == V4L2_FIELD_ANY || fmt->field == V4L2_FIELD_ALTERNATE)
 		fmt->field = V4L2_FIELD_NONE;
@@ -242,9 +242,9 @@ static int dcmipp_isp_enum_frame_size(st
 		return -EINVAL;
 
 	fse->min_width = DCMIPP_FRAME_MIN_WIDTH;
-	fse->max_width = DCMIPP_FRAME_MAX_WIDTH;
+	fse->max_width = DCMIPP_PIXEL_FRAME_MAX_WIDTH;
 	fse->min_height = DCMIPP_FRAME_MIN_HEIGHT;
-	fse->max_height = DCMIPP_FRAME_MAX_HEIGHT;
+	fse->max_height = DCMIPP_PIXEL_FRAME_MAX_HEIGHT;
 
 	return 0;
 }
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelcap.c
@@ -317,12 +317,12 @@ static int dcmipp_pixelcap_try_fmt_vid_c
 	in_w = format->width;
 	in_h = format->height;
 	format->width = clamp_t(u32, format->width, DCMIPP_FRAME_MIN_WIDTH,
-				DCMIPP_FRAME_MAX_WIDTH);
+				DCMIPP_PIXEL_FRAME_MAX_WIDTH);
 	format->width = round_up(format->width,
 				 1 << hdw_pixel_alignment(format->pixelformat));
 	format->height = clamp_t(u32, format->height,
 				 DCMIPP_FRAME_MIN_HEIGHT,
-				 DCMIPP_FRAME_MAX_HEIGHT);
+				 DCMIPP_PIXEL_FRAME_MAX_HEIGHT);
 	if (format->width != in_w || format->height != in_h)
 		dev_dbg(vcap->dev,
 			"resolution updated: %dx%d -> %dx%d\n",
@@ -435,9 +435,9 @@ static int dcmipp_pixelcap_enum_framesiz
 
 	fsize->type = V4L2_FRMSIZE_TYPE_CONTINUOUS;
 	fsize->stepwise.min_width = DCMIPP_FRAME_MIN_WIDTH;
-	fsize->stepwise.max_width = DCMIPP_FRAME_MAX_WIDTH;
+	fsize->stepwise.max_width = DCMIPP_PIXEL_FRAME_MAX_WIDTH;
 	fsize->stepwise.min_height = DCMIPP_FRAME_MIN_HEIGHT;
-	fsize->stepwise.max_height = DCMIPP_FRAME_MAX_HEIGHT;
+	fsize->stepwise.max_height = DCMIPP_PIXEL_FRAME_MAX_HEIGHT;
 	fsize->stepwise.step_width = 1;
 	fsize->stepwise.step_height = 1;
 
--- a/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
+++ b/drivers/media/platform/st/stm32/stm32-dcmipp/dcmipp-pixelproc.c
@@ -530,9 +530,9 @@ static void dcmipp_pixelproc_adjust_fmt(
 		fmt->code = PIXELPROC_MEDIA_BUS_FMT_DEFAULT;
 
 	fmt->width = clamp_t(u32, fmt->width, DCMIPP_FRAME_MIN_WIDTH,
-			     DCMIPP_FRAME_MAX_WIDTH);
+			     DCMIPP_PIXEL_FRAME_MAX_WIDTH);
 	fmt->height = clamp_t(u32, fmt->height, DCMIPP_FRAME_MIN_HEIGHT,
-			      DCMIPP_FRAME_MAX_HEIGHT);
+			      DCMIPP_PIXEL_FRAME_MAX_HEIGHT);
 
 	if (fmt->field == V4L2_FIELD_ANY || fmt->field == V4L2_FIELD_ALTERNATE)
 		fmt->field = V4L2_FIELD_NONE;
@@ -618,9 +618,9 @@ static int dcmipp_pixelproc_enum_frame_s
 		return -EINVAL;
 
 	fse->min_width = DCMIPP_FRAME_MIN_WIDTH;
-	fse->max_width = DCMIPP_FRAME_MAX_WIDTH;
+	fse->max_width = DCMIPP_PIXEL_FRAME_MAX_WIDTH;
 	fse->min_height = DCMIPP_FRAME_MIN_HEIGHT;
-	fse->max_height = DCMIPP_FRAME_MAX_HEIGHT;
+	fse->max_height = DCMIPP_PIXEL_FRAME_MAX_HEIGHT;
 
 	return 0;
 }
@@ -818,8 +818,8 @@ dcmipp_pixelproc_enum_frame_interval(str
 
 	if (fie->pad > 1 ||
 	    fie->index >= (IS_SRC(fie->pad) ? ARRAY_SIZE(dcmipp_frates) : 1) ||
-	    fie->width > DCMIPP_FRAME_MAX_WIDTH ||
-	    fie->height > DCMIPP_FRAME_MAX_HEIGHT)
+	    fie->width > DCMIPP_PIXEL_FRAME_MAX_WIDTH ||
+	    fie->height > DCMIPP_PIXEL_FRAME_MAX_HEIGHT)
 		return -EINVAL;
 
 	if (IS_SINK(fie->pad)) {
